docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md

```diff
@@ -0,0 +1,535 @@
+
+model.initialize(init=init.Xavier(), ctx=ctx)
+model.hybridize()
+
+# Attach a pre-trained glove word vector to the embedding layer
+model.embedding_layer.weight.set_data(vocab.embedding.idx_to_vec)
+# fixed the layer
+model.embedding_layer.collect_params().setattr('grad_req', 'null')
+```
+
+Using r attention can improve the representation of sentences with different semantics, but if the value of each line in the attention matrix A (r-byn) is very close, that is, there is no difference between several attentions. In the subsequent M = AH, the resulting M will contain a lot of redundant information.
+So in order to solve this problem, we should try to force A to ensure that the value of each line has obvious differences, that is, try to satisfy the diversity of attention. Therefore, a penalty can be used to achieve this goal.
+$$ P = ||(AA^T-I)||_F^2 $$
+
+
+It can be seen from the above formula that if the value of each row of A is more similar, the result of P will be larger, and the value of A is less similar for each row, and P is smaller. This means that when the r-focused diversity of A is larger, the smaller P is. So by including this penalty item with the Loss of the model, you can try to ensure the diversity of A.
+
+```{.python .input}
+def evaluate(data_iter_valid, model, loss, ctx, penal_coeff=0.0, class_weight=None, loss_name='wsce'):
+    valid_loss = 0.
+    total_pred = []
+    total_true = []
+    n_batch = 0
+    for batch_x, batch_y in data_iter_valid:
+        batch_x = batch_x.as_in_context(ctx)
+        batch_y = batch_y.as_in_context(ctx)
+        batch_pred, att = model(batch_x)
+        if loss_name == 'sce':
+            l = loss(batch_pred, batch_y)
+        elif loss_name == 'wsce':
+            l = loss(batch_pred, batch_y, class_weight, class_weight.shape[0])
+
+        # penalty
+        temp = nd.batch_dot(att, nd.transpose(att, axes=(0, 2, 1))
+                            ) - nd.eye(att.shape[1], ctx=att.context)
+        l = l + penal_coeff * temp.norm(axis=(1, 2))
+        total_pred.extend(np.argmax(nd.softmax(batch_pred, axis=1).asnumpy(), axis=1).tolist())
+        total_true.extend(np.reshape(batch_y.asnumpy(), (-1,)).tolist())
+        n_batch += 1
+        valid_loss += l.mean().asscalar()
+
+    F1_valid = f1_score(np.array(total_true), np.array(total_pred), average='weighted')
+    acc_valid = accuracy_score(np.array(total_true), np.array(total_pred))
+    valid_loss /= n_batch
+
+    return F1_valid, acc_valid, valid_loss
+```
+
+```{.python .input}
+
+def train(data_iter_train, data_iter_valid, model, loss, trainer,
+          ctx, num_epochs, penal_coeff=0.0, clip=None,
+          class_weight=None, loss_name='wsce'):
+
+    print('Train on ', ctx)
+    for epoch in range(1, num_epochs + 1):
+        start = time.time()
+        train_loss = 0.
+        total_pred = []
+        total_true = []
+        n_batch = 0
+
+        for batch_x, batch_y in data_iter_train:
```

Comment:
There is quite a bit of copy & paste between `evaluate()` and `train()`, which should be consolidated. Note that you can use something like 
```
            with ExitStack() as stack:
                if is_train:
                    stack.enter_context(autograd.record())
```
to make sure you enter `autograd.record()` context only in training.

Reply:
Because there are gradient calculations, cropping, parameter updates, learning rate reduction, and output progress in training. But not in the evaluation. I followed the train and evaluation in the gluonbook and made a simple modification.

Reply:
yes, there are additional operations, but you can make it optional by making `is_train` a parameter. I understand that other notebooks don't do a good job in code sharing between training and evaluation, but that shouldn't prevent us from writing better code; we should set up a better standard.

Reply:
You are right, I have modified this problem to minimize the writing of duplicate code, but I don't understand ExitStack very well.